CIPs/cip-draft_poprf.md

```diff
@@ -0,0 +1,475 @@
+stored alongside the ciphertext.
+
+```typescript
+type PepperedDomain = {
+  name: "ODIS Peppered Domain"
+  version: 1
+  // Public key of a key-pair derived from the pepper.
+  publicKey: string
+  // Maximum number of requests that can be made against this domain.
+  cap?: number
+}
+
+type PepperedDomainOptions = {
+  // Nonce value to include in the signature message to make it unique.
+  nonce: number
+  // EIP-712 signature over the request.
+  signature: string
+}
+```
+
+**Phone number hashing:** The current ODIS rate limiting scheme could be implemented as a particular
+domain in the new API. A domain string and options could be written as follows:
+
+```typescript
+type PhoneNumberDomain = {
+  name: "ODIS Phone Number Domain"
+  version: 1
+}
+
+type PhoneNumberDomainOptions = {
+  // Celo account address. Query is charge against this account's quota.
+  account: string
+  // Signature from the account, or authorized signer for the account.
+  authorization: string
+  // Optional on-chain identifier. Unlocks additional quota if the
+  // account is verified as an owner of the identifier.
+  hashedPhoneNumber?: string
+}
+```
+
+**Smart contract access controlled:** Useful for expanding the set of use cases to community derived
+ones. The address of a smart contract implementing the following interface will be a parameter in
+the domain.
+
+```solidity
+interface OdisAccessController {
+  function shouldAllowOdisRequest(bytes calldata domainData, bytes calldata authorization) external view returns(bool);
+}
+```
+
+For example, this contract could check that a user owns a certain NFT or has burned a fee before
+being allowed access to the signature. The signature would be applied to the domain and the returned
+value from the contract function.
+
+```typescript
+type SmartContractDomain = {
+  name: "ODIS Smart Contract Domain"
+  version: 1
+  // Address of the contract that controls access to this domain.
+  // Call will occur as a view call at the chain head upon receiving
+  // the request to determine rate limiting.
+  contractAddress: string
+  // Specified to ensure an unambiguous smart contract is resolved.
+  // An ODIS given service is likely to support only a single chain.
+  chainId: number
+  // Arbitrary bytes to be included in the smart contract call for
+  // use by the contract logic.
+  domainData?: string
+}
+
+type SmartContractDomainOptions = {
+  // Optional parameter to avoid issues if the ODIS service node is behind.
+  // When specified, the authorization check should not occur against a block
+  // prior to the specified number.
+  notBeforeBlock?: number
+  // Arbitrary bytes to be included in the smart contract call for use by the
+  // contract logic. Intended to be a signature or other authorization string.
+  authorization: string
+}
+```
+
+### Domain serialization
+
+Because any variation in the domain string, by design, results in an unpredictable change in the
```

Comment:
Maybe this hasn't been thought out yet (or I misunderstand ODIS), but how exactly would a new domain name correspond to new/different output? My understanding is that ODIS generates a pepper to hash phone numbers from a set of keys created during the DKG ceremony - is this pepper what essentially distinguishes the output of different domains? In that case, wouldn't there need to be a new DKG ceremony for each domain, so how would a completely new domain `name` be handled (since there may not be a set of keys/already-performed DKG corresponding to it)?

Reply:
Seems to be clarified in the `HTTP Transport` section

Reply:
A single key is used for all domains. Not sure what in the `HTTP Transport` section answers this 😅